radix-engine/src/engine/process.rs

```diff
@@ -208,36 +208,66 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
         }
     }
 
+    fn new_bucket_id(&mut self) -> Result<BucketId, RuntimeError> {
+        if self.depth == 0 {
+            self.id_allocator
+                .new_bucket_id()
+                .map_err(RuntimeError::IdAllocatorError)
+        } else {
+            Ok(self.track.new_bucket_id())
+        }
+    }
+
+    fn new_proof_id(&mut self) -> Result<ProofId, RuntimeError> {
+        if self.depth == 0 {
+            self.id_allocator
+                .new_proof_id()
+                .map_err(RuntimeError::IdAllocatorError)
+        } else {
+            Ok(self.track.new_proof_id())
+        }
+    }
+
     // (Transaction ONLY) Takes resource from worktop and returns a bucket.
     pub fn take_from_worktop(
         &mut self,
-        resource_spec: ResourceSpecification,
+        resource: ResourceSpecifier,
     ) -> Result<ValidatedData, RuntimeError> {
-        re_debug!(
-            self,
-            "(Transaction) Taking from worktop: {:?}",
-            resource_spec
-        );
-        let resource_def_id = resource_spec.resource_def_id();
-        let new_bucket_id = self
-            .id_allocator
-            .new_bucket_id()
-            .map_err(RuntimeError::IdAllocatorError)?;
-        let bucket = match resource_spec {
-            ResourceSpecification::Fungible { amount, .. } => {
-                self.worktop.take(amount, resource_def_id)
-            }
-            ResourceSpecification::NonFungible { ids, .. } => {
-                self.worktop.take_non_fungibles(&ids, resource_def_id)
+        re_debug!(self, "(Transaction) Taking from worktop: {:?}", resource);
+        let new_bucket_id = self.new_bucket_id()?;
+        let bucket = match resource {
+            ResourceSpecifier::Amount(amount, resource_def_id) => self
```

Comment:
Let me try it.

Reply:
Actually, I will just remove this match logic and expose three methods. There is no point merging three transaction instructions (`take_from_worktop`, `take_non_fungibles_from_worktop` and `take_all_from_worktop`) into one and splitting it in the `Process`.

This will further reduce `ResourceSpecifier` to resim/tx_builder, to be aligned with our decision on amount types.